diff --git a/js/icon_select_frontend.js b/js/icon_select_frontend.js
index e0f0ece..1fd5ef3 100644
--- a/js/icon_select_frontend.js
+++ b/js/icon_select_frontend.js
@@ -19,6 +19,13 @@
         }
         var svg = xhr.responseXML.documentElement;
         svg = document.importNode(svg, true);
+        // Ensures that SVG groups appended to end of page have unique ids.
+        var gElements = svg.querySelectorAll('g');
+        if (gElements) {
+          gElements.forEach(function (element, index) {
+            if (element.hasAttribute('id')) {
+              element.id = element.id + '_' + Date.now();
+            }
+          });
+        }
         document.body.appendChild(svg);
       };
 
